src/NuGet.Clients/NuGet.VisualStudio.Common/ErrorListTableDataSource.cs

```diff
@@ -205,23 +204,23 @@ private void EnsureInitialized()
             if (!_initialized)
             {
                 // Double check around locking since this is called often.
-                lock (_initLockObj)
-                {
-                    if (!_initialized)
+                LazyInitializer.EnsureInitialized(ref _tableManager,
```

Comment:
One of the remarks in https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1, scares me. 

In particular: 

```
In the event that multiple threads access this method concurrently, multiple instances of T may be created, but only one will be stored into target. In such an occurrence, this method will not dispose of the objects that were not stored. If such objects must be disposed, it is up to the caller to determine if an object was not used and to then dispose of the object appropriately.

```

I don't know if the initialization below can handle that. 

Now the likelihood of concurrent initialization like is super low, but it does give me a pause.

Reply:
Maybe Lazy<T> is better choice here, as it is guaranteed to be thread-safe, and single-initialized.

> By default, Lazy<T> objects are thread-safe. [...] In multi-threaded scenarios, the first thread to access the Value property of a thread-safe Lazy<T> object initializes it for all subsequent accesses on all threads, and all threads share the same data. 

https://docs.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization

Reply:
A lazy would deadlock if called from a background thread. Think we already have that problem with extensiblesourceprovider. 

I'll dig more but I think the LazyInitializer is deadlock prone, and I now think it might be a good idea to say "don't use it" in VS codepaths.

Reply:
> One of the remarks in https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1, scares me.
> 
> In particular:
> 
> ```
> In the event that multiple threads access this method concurrently, multiple instances of T may be created, but only one will be stored into target. In such an occurrence, this method will not dispose of the objects that were not stored. If such objects must be disposed, it is up to the caller to determine if an object was not used and to then dispose of the object appropriately.
> ```
> 
> I don't know if the initialization below can handle that.

As per my understanding, the [`overloaded method of EnsureInitialized`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=netcore-3.1#System_Threading_LazyInitializer_EnsureInitialized__1___0__System_Boolean__System_Object__System_Func___0__) invoked here is thread safe.

Reply:
@nkolev92 - I didn't know that Lazy<T> would deadlock... Did you encounter it yourself? I am asking because we are extensively use for singletons in VS code base 😐

Maybe AsyncLazy<T> would be better? You can also pass JTF to it...

Reply:
If a lazy has any thread affinity it can easily lead to a deadlock.
Ex: The lazy itself is called on the UI thread, but the code inside the tries to switch to background/ui thread at some point. See https://github.com/NuGet/NuGet.Client/pull/2919. 

I'll try to come back to this tomorrow.

Reply:
Thanks, @nkolev92 -- yeah, it seems that `AsyncLazy<T>` is preferred over `Lazy<T>`

Reply:
@nkolev92 there is no `AsyncLazy` version of this `LazyInitializer.EnsureInitialized`. Do we just abandon this PR?

Reply:
```cs
namespace Microsoft.VisualStudio.Threading
{
    //
    // Summary:
    //     Lazily executes a delegate that has some side effect (typically initializing
    //     something) such that the delegate runs at most once.
    public class AsyncLazyInitializer

```

Visual Studio threading has an AsyncLazyInitializer. 

LazyInitializer will deadlock when the blocked thread is a UI thread and the initialization inside needs the UI thread.

Honestly, while this was a great exercise, I'm not sure there'll be an actual benefit. Don't fix what's not broken.